accepted/2021/InboxSourceGenerators.md

```diff
@@ -0,0 +1,187 @@
+
+Due to the limited servicing agility for code-generated code, we should be very careful about the complexity of the code which is generated.  Where possible we should limit generated code to "glue code", only that which must be specific to the user's assembly or types.  We should try to put more complex code inside the framework or library itself, even if it means exposing public API that is specifically for source generated code to call.
+
+Inbox source generators will be serviced in their side-by-side location.  Should a bug exist in a osurce generator we will evaluate the bug against the bar for each framework it exists and patch them independently.  In this way the servicing process for these source generators will match the .NET runtime.
+
+**Open issue:** should we do anything to make it easier to identify assemblies which might need an update?  We can consider some case studies.  
+
+If a specific compiler version has a code-gen bug it is not clear from the final assembly if that bug is present, unless you directly inspect the IL/bytecode for the issue.  The compiler version is embedded in the PDB of the assembly which could be used to identify assemblies with the issue.  There is no notification at design time or runtime of any issue with the compiler, the user needs to subscribe to updates to understand when an update is needed.
+
+If a runtime assembly has a bug, we have multiple avenues of detection.  We could examine the assembly metadata files contained in the application layout, including for self-contained applications.  The runtime will also write "breadcrumb" files on a machine when code is executed that is annotated for servicing that help identify that a specific assembly or package version has been executed on that machine.  Today we use these breadcrumbs to help target updates to a machine.  Those updates can install in a well-known location and be preferred over the application's copy.
+
+For means of detection we have:
+1. Manual
+2. PDB indicator: source generator assembly version, file version, hash
+3. Assembly metadata in consuming user project: GeneratedCodeAttribute("System.Foo.Generator", "5.0.0+cf258a14b70ad9069470a108f13765e0e5988f51")
+4. Runtime breadcrumbs
+
+For means of update we have:
+1. No notification, user needs to subscribe to servicing updates.
+2. Design time notification.
+3. Killbit in the runtime to prevent execution of user assemblies that contain vulnerable source generated code.
```

Comment:
If we want to implement this one it would need the GeneratedCodeAttribute in the assembly metadata, right? PDB is generally not available to the runtime.

Reply:
Yeah.  We could impose this requirement ourselves in our source generators.  I think we need to understand if it's worth it to add a stamp to the runtime assembly or not.

Reply:
Discussion is happening here https://github.com/dotnet/designs/pull/181#discussion_r588772008